LibraBFT/Impl/Consensus/Liveness/Properties/ProposerElection.agda

```diff
@@ -0,0 +1,64 @@
+{- Byzantine Fault Tolerant Consensus Verification in Agda, version 0.9.
+
+   Copyright (c) 2021, Oracle and/or its affiliates.
+   Licensed under the Universal Permissive License v 1.0 as shown at https://opensource.oracle.com/licenses/upl
+-}
+
+open import Optics.All
+open import LibraBFT.Base.Types
+open import LibraBFT.ImplShared.Base.Types
+open import LibraBFT.ImplShared.Consensus.Types
+open import LibraBFT.ImplShared.Util.Util
+open import LibraBFT.Impl.Consensus.Liveness.ProposerElection
+open import LibraBFT.Lemmas
+open import LibraBFT.Prelude
+
+module LibraBFT.Impl.Consensus.Liveness.Properties.ProposerElection where
+
+-- TUTORIAL
+module isValidProposalMSpec (b : Block) where
+
+  pe      = _^∙ lProposerElection
+  mAuthor = b ^∙ bAuthor
+  round   = b ^∙ bRound
+
+  contract
+    : ∀ pre Post
+      → (mAuthor ≡ nothing → Post (Left ProposalDoesNotHaveAnAuthor) pre [])
+      → (Maybe-Any (getValidProposer (pe pre) round ≢_) mAuthor
+         → Post (Left ProposerForBlockIsNotValidForThisRound) pre [])
+      → (Maybe-Any (getValidProposer (pe pre) round ≡_) mAuthor
+         → Post (Right unit) pre [])
+      → LBFT-weakestPre (isValidProposalM b) Post pre
+  -- 1. `isValidProposalM` begins with `caseMM`, so we must provide two cases:
+  --    one where `b ^∙ bAuthor` is `nothing` and one where it is `just`
+  --    something
+  -- 2. When it is nothing, we appeal to the assumed proof
+  proj₁ (contract pre Post pfNone pf≢ pfOk) mAuthor≡nothing = pfNone mAuthor≡nothing
+  -- 3. When it is something, we step into `isValidProposerM`. This means:
+  --    - we use the proposer election of the round manager (`pe` and `pe≡`)
```

Comment:
The description is out of date, referring to names in the commented-out previous proof.  Maybe just add a `TODO-1` comment to update it, rather than spending time on it now.

Reply:
`pe` and `pe≡` do not refer to the commented-out previous proof, but rather to an intermediate step in the current proof. However, I see that the commented out section was still referring to `prfT` and `prfF` -- this has been corrected.